packages/starlight/user-components/rehype-file-tree.ts

```diff
@@ -0,0 +1,245 @@
+import { AstroError } from 'astro/errors';
+import type { Element, Text } from 'hast';
+import { type Child, h, s } from 'hastscript';
+import { select, selectAll } from 'hast-util-select';
+import { fromHtml } from 'hast-util-from-html';
+import { toString } from 'hast-util-to-string';
+import { rehype } from 'rehype';
+import { CONTINUE, SKIP, visit } from 'unist-util-visit';
+import { Icons } from '../components/Icons';
+import { definitions } from './file-tree-icons';
+
+declare module 'vfile' {
+	interface DataMap {
+		directoryLabel: string;
+	}
+}
+
+const folderIcon = makeSVGIcon(Icons['seti:folder']);
+const defaultFileIcon = makeSVGIcon(Icons['seti:default']);
+
+/**
+ * Process the HTML for a file tree to create the necessary markup for each file and directory
+ * including icons.
+ * @param html Inner HTML passed to the `<FileTree>` component.
+ * @param directoryLabel The localized label for a directory.
+ * @returns The processed HTML for the file tree.
+ */
+export function processFileTree(html: string, directoryLabel: string) {
+	const file = fileTreeProcessor.processSync({ data: { directoryLabel }, value: html });
+
+	return file.toString();
+}
+
+/** Rehype processor to extract file tree data and turn each entry into its associated markup. */
+const fileTreeProcessor = rehype().use(function fileTree() {
```

Comment:
Ah, just noticed one thing while referencing this PR during work on `<Steps>`! Should `rehype` be run in “fragment” mode here?

```suggestion
const fileTreeProcessor = rehype()
	.data('settings', { fragment: true })
	.use(function fileTree() {
```

This will change the shape of the parsed HTML (no `<body>` etc.), but is what our other component rehype pipelines do (e.g. tabs)

Reply:
Great catch, did not even notice that, just [pushed](https://github.com/withastro/starlight/pull/1308/commits/d9dfc9f09afeef757a4464e27077b08769802e18) a new version fixing that, and it even avoids relying on `selectAll` :tada:

The change seems huge but it's mostly due to indentation changes, the actual change is just a few lines.

Reply:
Nice! LGTM 🙌